include/ignition/rendering/base/BaseGeometry.hh

```diff
@@ -89,12 +97,37 @@ namespace ignition
       if (material) this->SetMaterial(material, _unique);
     }
 
+    //////////////////////////////////////////////////
+    template <class T>
+    GeometryPtr BaseGeometry<T>::Clone() const
+    {
+      if (!this->Scene())
+      {
+        ignerr << "Cloning a geometry failed because the geometry to be "
+          << "cloned does not belong to a scene.\n";
+        return nullptr;
+      }
+      else if (this->meshName.empty())
+      {
+        ignerr << "Cloning a geometry failed because the name of the mesh is "
+          << "missing.\n";
+        return nullptr;
+      }
```

Comment:
so far I think all geometries have a name. I don't know if an empty geometry is useful though. I think we can output error msg like you have now. 

On a related note, I think we should propbably store a copy of the full MeshDescriptor object instead of just the mesh name

Reply:
> I think we can output error msg like you have now

Should we also return `nullptr` if the mesh is empty, or just print the message and then continue the attempt to clone the geometry?

> On a related note, I think we should probably store a copy of the full MeshDescriptor object instead of just the mesh name

Is there a particular advantage to storing the full `MeshDescriptor` object? My concern is that the `MeshDescriptor` contains a [pointer to the mesh](https://github.com/ignitionrobotics/ign-rendering/blob/843888ad67f71f9296bb407f651133520bb822a5/include/ignition/rendering/MeshDescriptor.hh#L65), so I want to make sure that the mesh being pointed to by the `MeshDescriptor` isn't going to be deleted by another part of the program while it's still being used by a `MeshDescriptor`.

Reply:
I think we can return null if mesh empty

The MeshDescriptor contains other info including `submeshName`, and `centerSubmesh`.  It's possible that a rendering::Mesh object does not contain a full mesh but just a single submesh so when cloning the Mesh object, we don't want to copy the full mesh. e.g. a dae file can contain a wheeled robot mesh consisting of chassis and wheels and we usually create separate visuals for moving parts by using the [<submesh>](http://sdformat.org/spec?ver=1.8&elem=geometry#mesh_submesh) sdf element. 

We can probably leave out the pointer to the common::Mesh object.

Reply:
Sounds good :+1: I believe the functionality for storing a copy of the `MeshDescriptor` (excluding the `common::Mesh` pointer) is now complete: see https://github.com/ignitionrobotics/ign-rendering/commit/f5b07ea19707d9f2ba97565f5d7d5140d496f238